src/colorizer/FrameCache.ts

```diff
@@ -69,78 +75,66 @@ export default class FrameCache {
     this.setFirst(entry);
   }
 
-  public get length(): number {
-    return this.data.length;
+  public get size(): number {
+    return this.currentSize;
   }
 
-  /** Inserts a frame into the cache at `index` */
-  public insert(index: number, frame: Texture): void {
-    if (index >= this.data.length) {
-      return;
+  /**
+   * Inserts a value into the cache with the key. If the key already exists, the value is replaced.
+   * If the cache is full, the least recently used entry is evicted and its `dispose()` function is
+   * called if it exists.
+   *
+   * Throws an error if the `size` is greater than the cache's `maxSize`, as defined in the
+   * constructor.
+   */
+  public insert(key: string | number, value: E, size: number = 1): void {
+    key = key.toString();
+    const currentEntry = this.data.get(key);
+
+    if (size > this.maxSize) {
+      throw new Error(`Attempted to insert a value of size ${size} into a cache with maxSize ${this.maxSize}`);
     }
 
-    const currentEntry = this.data[index];
-    if (currentEntry !== null) {
-      currentEntry.frame = frame;
+    if (currentEntry !== null && currentEntry !== undefined) {
+      // NOTE: This assumes that the value's size HAS NOT CHANGED. If it has,
+      // the size of the cache will be incorrect and the old value will not be disposed of.
+      // TODO: Throw an error if the size is different?
+      currentEntry.value = value;
       this.moveToFirst(currentEntry);
     } else {
-      const newEntry = { frame, index, prev: null, next: null };
-      this.data[index] = newEntry;
-      this.setFirst(newEntry);
+      const newEntry = { value, key, prev: null, next: null, size: size };
+      this.data.set(key, newEntry);
 
-      if (this.numItems >= this.maxSize) {
+      this.currentSize += size;
+      while (this.currentSize > this.maxSize && this.last) {
         this.evictLast();
-      } else {
-        if (!this.last) {
-          this.last = newEntry;
-        }
-        this.numItems++;
+      }
```

Comment:
sanity check: does evictLast update currentSize?

Reply:
It does!